django/contrib/sessions/backends/base.py

```diff
@@ -375,3 +386,353 @@ def clear_expired(cls):
         a built-in expiration mechanism, it should be a no-op.
         """
         raise NotImplementedError('This backend does not support clear_expired().')
+
+
+class HashingSessionBase(SessionBase):
+    """
+    Base class for all Session classes since introducing
+    session-key hashing.
+    """
+    _algorithm = getattr(hashlib, SESSION_HASHING_ALGORITHM)
+
+    def _get_frontend_key(self):
+        return super()._get_session_key()
+
+    def _set_frontend_key(self, key):
+        return super()._set_session_key(key)
+
+    frontend_key = property(_get_frontend_key)
+
+    _frontend_key = property(_get_frontend_key, _set_frontend_key)
+
+    @classmethod
+    def get_serializer(cls):
+        return import_string(settings.SESSION_SERIALIZER)
+
+    @classmethod
+    def get_key_salt(cls):
+        return 'django.contrib.sessions.' + cls.__qualname__
+
+    @classmethod
+    def _encode(cls, session_dict):
+        "Return the given session dictionary serialized and encoded as a string."
+        return signing.dumps(
+            session_dict, salt=cls.get_key_salt(), serializer=cls.get_serializer(),
+            compress=True,
+        )
+
+    def encode(self, session_dict):
+        # RemovedInDjango40Warning: use _encode classmethod
+        return self._encode(session_dict)
+
+    @classmethod
+    def __hash(cls, value):
```

Comment:
I don't understand why we have methods with a double underscores prefix which are copies from `SessionBase`, e.g. `__hash()`, `__legacy_encode()`, `__legacy_decode()` :thinking: